src/rrdp.rs

```diff
@@ -1,456 +1,962 @@
+        let json = serde_json::to_string_pretty(&self)?;
 
-impl From<XmlDecodeError> for DeltaParserError {
-    fn from(err: XmlDecodeError) -> Self {
-        DeltaParserError::Xml(err)
+        file_ops::write_buf(path, json.as_bytes()).with_context(|| "Could not save state.")
     }
-}
 
-impl<'a> DeltaParser<'a> {
-    fn new(snapshot: &'a mut SnapshotParser) -> Self {
-        DeltaParser {
-            meta: None,
-            snapshot,
-        }
+    pub fn session_id(&self) -> Uuid {
+        self.session_id
     }
 
-    fn is_applicable(&self) -> Result<bool, DeltaParserError> {
-        let self_meta = self.meta.as_ref().ok_or(
-            DeltaParserError::InternalErrorMissingDeltaMeta)?;
+    pub fn serial(&self) -> u64 {
+        self.serial
+    }
+
+    pub fn elements(&self) -> impl Iterator<Item = &CurrentObject> {
+        self.current_objects.objects()
+    }
 
-        let snapshot_meta = self.snapshot.meta.as_ref().ok_or(
-            DeltaParserError::InternalErrorMissingSnapshotMeta)?;
+    /// Writes the notification file to disk. Will first write to a
+    /// temporary file and then rename it to avoid serving partially
+    /// written files.
+    pub fn write_notification(&self) -> Result<()> {
+        let tmp_path = self.rrdp_dir.join(".notification.xml");
+        let final_path = self.rrdp_dir.join("notification.xml");
 
-        // info!("DeltaParser::is_applicable(): delta session_id={} snapshot session_id={}", &self_meta.session_id, &snapshot_meta.session_id);
-        // info!("DeltaParser::is_applicable(): delta serial={} snapshot serial={}", &self_meta.serial, &snapshot_meta.serial);
-        if snapshot_meta.session_id != self_meta.session_id {
-            Err(DeltaParserError::MismatchedSessionId(self_meta.session_id))
-        } else {
-            Ok(self_meta.serial > snapshot_meta.serial)
+        info!("Updating notification file at {:?}", final_path);
+
+        let notification = self.make_notification_file()?;
+
+        let mut bytes: Vec<u8> = vec![];
+        let mut writer = rpki::xml::encode::Writer::new_with_indent(&mut bytes);
+        notification.to_xml(&mut writer)?;
+
+        file_ops::write_buf(&tmp_path, &bytes)
+            .with_context(|| "Could not write temporary notification file")?;
+
+        fs::rename(tmp_path, final_path)
+            .with_context(|| "Could not rename tmp notification file to real notification file")?;
+
+        Ok(())
+    }
+
+    fn make_notification_file(&self) -> Result<NotificationFile> {
+        let base_uri = self
+            .notification_uri
+            .parent()
+            .ok_or_else(|| anyhow!("Notification URI does not have a parent?!"))?;
+
+        let rel_path_snapshot = Self::rel_path_snapshot(self.session_id(), self.serial());
+
+        let snapshot_uri = base_uri
+            .join(rel_path_snapshot.as_bytes())
+            .with_context(|| "Could not derive snapshot URI")?;
```

Comment:
Is derive the right word here? Should the path be mentioned in the error context?

Reply:
can't really happen, so removed it